### ISSUE TITLE
Arity confusion in decreases clauses

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#1901]
```
let rec g (a:nat) (x:list nat)
  : Tot (nat -> Tot nat (decreases x))
        (decreases x)
  = fun s ->
    match x with
    | [] -> s
    | hd::tl -> g a tl s
```

This function seems to need to be annotated with decreases clauses in two places. Removing either of the annotations causes verification to fail. The problem seems to have something to do with confusing arities of the recursive function in TcTerm.guard_let_recs

**Comment by mtzguido** [comment]
The nesting of `Tm_arrow` nodes is to blame. Potential quick-and-dirty fix 6c904617d9a3ef6a904904a9798e90bde3d8ee0d. It passes CI, except that it caused all the examples in #379 to be accepted (which I think is good!).

**Comment by mtzguido** [comment]
The original example here now passes, but the one in #1902 does not.

The fix I proposed still works and accepts both, but could cause a regression due to the arity check for #1017, which we've been discussing separately. See b0ad88e0639d066f02676f9e5f95e4322acc57cf and previous commits.
